File: cppVS/Strategy/ShapeTest.cpp
#include "pch.h"

#include "Shape.h"
#include "MockCanvas.h"
#include "Type.h"

TEST_CASE("A Shape can be created")
{
    Shape{ Type::Triangle };
}

TEST_CASE("A Shape can be drawn")
{
    const Shape s(Type::Triangle);
    MockCanvas c;
    ALLOW_CALL(c, DrawVertex(ANY(Point)));

    s.Draw(c);
}

TEST_CASE("A triangular shape draws correctly")
{
    const Shape s(Type::Triangle);
    MockCanvas c;

    REQUIRE_CALL(c, DrawVertex(Point{ 0, 0 })).TIMES(1);
    REQUIRE_CALL(c, DrawVertex(Point{ 2, 0 })).TIMES(1);
    REQUIRE_CALL(c, DrawVertex(Point{ 1, 2 })).TIMES(1);

    s.Draw(c);
}

TEST_CASE("A square shape draws correctly")
{
    const Shape s(Type::Square);
    MockCanvas c;

    REQUIRE_CALL(c, DrawVertex(Point{ 0, 0 })).TIMES(1);
    REQUIRE_CALL(c, DrawVertex(Point{ 2, 0 })).TIMES(1);
    REQUIRE_CALL(c, DrawVertex(Point{ 2, 2 })).TIMES(1);
    REQUIRE_CALL(c, DrawVertex(Point{ 0, 2 })).TIMES(1);

    s.Draw(c);
}

TEST_CASE("A pentagon shape draws correctly")
{
    const Shape s(Type::Pentagon);
    MockCanvas c;

    REQUIRE_CALL(c, DrawVertex(Point{ 0, 0 })).TIMES(1);
    REQUIRE_CALL(c, DrawVertex(Point{ 2, 0 })).TIMES(1);
    REQUIRE_CALL(c, DrawVertex(Point{ 2, 2 })).TIMES(1);
    REQUIRE_CALL(c, DrawVertex(Point{ 1, 3 })).TIMES(1);
    REQUIRE_CALL(c, DrawVertex(Point{ 0, 2 })).TIMES(1);

    s.Draw(c);
}

TEST_CASE("A hexagon shape draws correctly")
{
    const Shape s(Type::Hexagon);
    MockCanvas c;

    REQUIRE_CALL(c, DrawVertex(Point{ 0, 0 })).TIMES(1);
    REQUIRE_CALL(c, DrawVertex(Point{ 2, 0 })).TIMES(1);
    REQUIRE_CALL(c, DrawVertex(Point{ 3, 1 })).TIMES(1);
    REQUIRE_CALL(c, DrawVertex(Point{ 2, 2 })).TIMES(1);
    REQUIRE_CALL(c, DrawVertex(Point{ 0, 2 })).TIMES(1);
    REQUIRE_CALL(c, DrawVertex(Point{ -1, 1 })).TIMES(1);

    s.Draw(c);
}

File: CSharp/StrategyCSharp/ShapeTest.cs

using Moq;
using NUnit.Framework;

namespace StrategyCSharp
{
    class ShapeTest
    {
        [Test]
        public void AShapeCanBeCreated()
        {
            Shape s = new Shape(Type.Triangle);
        }

        [Test]
        public void AShapeCanBeDrawn()
        {
            Shape s = new Shape(Type.Triangle);
            var c = new Mock<ICanvas>();

            s.Draw(c.Object);
        }

        [Test]
        public void ATriangularShapeDrawsCorrectly()
        {
            Shape s = new Shape(Type.Triangle);
            var c = new Mock<ICanvas>();

            s.Draw(c.Object);

            c.Verify(mock => mock.DrawVertex(new Point(0, 0)), Times.Once());
            c.Verify(mock => mock.DrawVertex(new Point(2, 0)), Times.Once());
            c.Verify(mock => mock.DrawVertex(new Point(1, 2)), Times.Once());
        }

        [Test]
        public void ASquareShapeDrawsCorrectly()
        {
            Shape s = new Shape(Type.Square);
            var c = new Mock<ICanvas>();

            s.Draw(c.Object);

            c.Verify(mock => mock.DrawVertex(new Point(0, 0)), Times.Once());
            c.Verify(mock => mock.DrawVertex(new Point(2, 0)), Times.Once());
            c.Verify(mock => mock.DrawVertex(new Point(2, 2)), Times.Once());
            c.Verify(mock => mock.DrawVertex(new Point(0, 2)), Times.Once());
        }

        [Test]
        public void APentagonShapeDrawsCorrectly()
        {
            Shape s = new Shape(Type.Pentagon);
            var c = new Mock<ICanvas>();

            s.Draw(c.Object);

            c.Verify(mock => mock.DrawVertex(new Point(0, 0)), Times.Once());
            c.Verify(mock => mock.DrawVertex(new Point(2, 0)), Times.Once());
            c.Verify(mock => mock.DrawVertex(new Point(2, 2)), Times.Once());
            c.Verify(mock => mock.DrawVertex(new Point(1, 3)), Times.Once());
            c.Verify(mock => mock.DrawVertex(new Point(0, 2)), Times.Once());
        }

        [Test]
        public void AHexagonShapeDrawsCorrectly()
        {
            Shape s = new Shape(Type.Hexagon);
            var c = new Mock<ICanvas>();

            s.Draw(c.Object);

            c.Verify(mock => mock.DrawVertex(new Point(0, 0)), Times.Once());
            c.Verify(mock => mock.DrawVertex(new Point(2, 0)), Times.Once());
            c.Verify(mock => mock.DrawVertex(new Point(3, 1)), Times.Once());
            c.Verify(mock => mock.DrawVertex(new Point(2, 2)), Times.Once());
            c.Verify(mock => mock.DrawVertex(new Point(0, 2)), Times.Once());
            c.Verify(mock => mock.DrawVertex(new Point(-1, 1)), Times.Once());
        }
    }
}
